dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/datafile/page/WholeDatasetDownloadLogger.java

```diff
@@ -0,0 +1,117 @@
+import edu.harvard.iq.dataverse.persistence.dataset.Dataset;
+import edu.harvard.iq.dataverse.persistence.dataset.DatasetVersion;
+
+import javax.ejb.Stateless;
+import javax.inject.Inject;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import java.util.stream.Collectors;
+
+import static java.util.stream.Collectors.toList;
+import static java.util.stream.Collectors.toSet;
+
+@Stateless
+public class WholeDatasetDownloadLogger {
+    private static final Logger logger = Logger.getLogger(WholeDatasetDownloadLogger.class.getSimpleName());
+
+    private DownloadDatasetLogService downloadDatasetLogService;
+
+    // -------------------- CONSTRUCTORS --------------------
+
+    @Deprecated
+    public WholeDatasetDownloadLogger() { }
+
+    @Inject
+    public WholeDatasetDownloadLogger(DownloadDatasetLogService downloadDatasetLogService) {
+        this.downloadDatasetLogService = downloadDatasetLogService;
+    }
+
+    // -------------------- LOGIC --------------------
+
+    public void incrementLogIfDownloadingWholeDataset(List<DataFile> filesToDownload) {
+        if (Lists.isEmpty(filesToDownload)) {
+            logger.warning("Empty or null file metadata list.");
+            return;
+        }
+
+        long startTime = System.currentTimeMillis();
+        Map<Dataset, List<DataFile>> filesByDataset = groupFilesByDatasetsAndUpdateLogIfNeeded(filesToDownload);
+        long totalTime = System.currentTimeMillis() - startTime;
+
+        logger.log(Level.FINE, "Checking for whole dataset download in datasets [{0}] took: {1} ms",
+                new Object[] { extractDatasetIds(filesByDataset), totalTime });
+    }
+
+    // -------------------- PRIVATE --------------------
+
+    private Map<Dataset, List<DataFile>> groupFilesByDatasetsAndUpdateLogIfNeeded(List<DataFile> filesToDownload) {
+        Map<Dataset, List<DataFile>> filesByDataset = filesToDownload.stream()
+                .filter(f -> f != null && f.getOwner() != null)
+                .collect(Collectors.groupingBy(DataFile::getOwner));
+
+        filesByDataset.forEach(this::updateLogOnWholeDatasetDownload);
+
+        return filesByDataset;
+    }
+
+    private void updateLogOnWholeDatasetDownload(Dataset dataset, List<DataFile> dataFiles) {
+        if (isUserDownloadingAllFilesFromAnyVersion(dataset, dataFiles)) {
+            downloadDatasetLogService.logWholeSetDownload(dataset.getId());
+        }
+    }
+
+    /**
+     * Here we are extracting sets of ids of files from all the versions that have no more files than
+     * list of files being downloaded and we check if any of these sets is wholly contained in the set of ids
+     * of files being downloaded
+     */
+    private boolean isUserDownloadingAllFilesFromAnyVersion(Dataset dataset, List<DataFile> dataFiles) {
```

Comment:
In tandem with comment on `createPossibleSetsOfFileIdsFromVersions`.

The way I'd imagine it work:
1. Check embargo on dataset (if there is one should we consider users permissions or just abort? - maybe discuss with @wfenrich )
2. Generate sets of filesids 
3. Find out if `wholedatasetdownload` is happening

Reply:
If I correctly understand, after removing check for embargo, this should behave as you've written.

Reply:
Yes, we have agreed on removing embargo checks while counting downloads. If downloads happens we don't bother checking for possible reasons why it did, we just count it.